wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -149,12 +147,30 @@ protected Builder() {
      * this message has no unknown fields.
      */
     public final ByteString buildUnknownFields() {
-      return unknownFieldsBuffer != null
-          ? unknownFieldsBuffer.clone().readByteString()
-          : ByteString.EMPTY;
+      if (unknownFieldsBuffer != null) {
+        // Reads and caches the unknown fields from the buffer.
+        unknownFieldsByteString = unknownFieldsBuffer.readByteString();
+        unknownFieldsBuffer = null;
```

Comment:
The reason for clearing unknownFieldBuffer is to ensure that it has no unreleased segments in the buffer which will cause memory leak when translating wire into objc.

The unknown fields are cached in unknownFieldsByteString. Next time when the caller adds an additional field, the method prepareForNewUnknownFields () will be called, which will read unknownFieldsByteString into the new Buffer again. We won't lose any unknown field.